### ISSUE TITLE
How to handle checkboxes?

### ISSUE THREAD

**Issue by marick** [boydm/phoenix_integration#33]
I have checkboxes that I am perhaps creating wrongly - if so, what would be the right way?

```elixir
    <%= for a <- @animals do %>
       <div class="field">
          <div class="ui checkbox">
            <%= checkbox(f, :animal_ids,
                  name: "#{input_name(:animals, :chosen_animal_ids)}[#{a.id}]") %>
```

In manual testing, the checkboxes work as expected. In particular, the checkboxes are unchecked when the form first shows. 

However, `fetch_form` shows both values starting as `"true"`:

```%{
  "chosen_animal_ids" => %{"237" => "true", "238" => "true"},
  "transaction_key" => "0f10e52f-abc4-49b5-9e05-14658295742f"
}
```

(Perhaps this is because the "real" checkbox input comes after the hidden one?)

Further, I can't seem to change the value the form delivers. For example, suppose I do this:

```elixir
       follow_form(has_animal_form,
         %{ animals:
            %{chosen_animal_ids: %{:"#{bossie.id}" => "true" }}})
```

What's delivered to the controller is this:

```elixir
%{
  "chosen_animal_ids" => %{"239" => "true", "240" => "true"},
  "transaction_key" => "a5743aa5-f1ce-42fe-ad12-1d44e97d86e1"
}
```

Is there a correct way to do this?

**Comment by marick** [comment]
I understand what's going on now, but a simple solution requires not using some `checkbox` options (most particularly `:hidden`). That makes me doubt that some sort of `make_sure_these_checkboxes_are_checked_and_no_others` function would be too fragile to include in a general library. 

I could write up a description of the issues and possible solutions, if you'd like to link it from somewhere in the documentation.

**Comment by marick** [comment]
My tentative function for making sure the right checkboxes are checked looks like this:

```
 defp checked_checkboxes(conn, field, keys) do
   should_be_true =
     Enum.map(keys, fn key -> to_string(key) |> String.to_atom end)

   all_keys =
     get_in(fetch_form(conn), [:inputs, :animals, field]) |> Map.keys

   pair = fn key ->
     value = Enum.member?(should_be_true, key) |> to_string
     {key, value}
   end

   %{field => Enum.map(all_keys, pair) |> Map.new}
```

Note all the specific values.

**Comment by boydm** [comment]
Wrapping my head around this. Can you write up a test that shows the failure?

**Comment by marick** [comment]
I've created pull request #34 that documents my understanding.

**Comment by marick** [comment]
Note: the failing tests have been marked with `@tag :skip`.

**Comment by marick** [comment]
It might be worth noting that my original idea was to take advantage of the `hidden_input: false` option to [checkbox](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#checkbox/3) to populate my form with fields like:

```
<input name="animals[chosen][1]" type="checkbox" value="name-of-animal">
```

... which would save me some code and a database round-trip.

**Comment by marick** [comment]
I've spiked up a solution that at least makes the code use hidden values for a checkbox's param when the test doesn't override them. 

```
      %{animals: %{chosen: checked}} =
        test_build_form_data(form, @form_action, %{})
      assert checked == %{"1": "false", "2": "false"}
```

It knows how to find hidden inputs that correspond to a `type="checkbox"`, update the checkbox's `value` to the hidden (default) value, and remove the hidden inputs from the form.

The spike is not exactly graceful:

https://github.com/marick/phoenix_integration/commit/118defc4ac588bfc52c101cd79f5ed92d981b6c9

... but it does I think have all the required pieces (except that I haven't handled the case where the checkbox starts out checked).

I have a rough glimmering of how to make it correctly handle things being set by the test, but I don't know if you want me to proceed on this project, or - if you do - *how* you want to proceed.

**Comment by marick** [comment]
(There are better ways to write the above code, but I *think* they would break special cases.)

**Comment by marick** [comment]
As I think about it, maybe the current code does the Right Thing™, which is to collapse multiple instances of the same `name` into a single one. That covers all the three cases that (I think) Phoenix handles:

* `name="something_that_is_not_an_array"`
* `name="user_wants_an_array_of_values[]"`
* `name="user_wants_a_map_of_keys_to_values[key]"`

It's easy for a test-writer to make use of that. For example, the test step that tricked me up could be:

```elixir
        follow_form(has_animal_form,
          %{animals:
            %{chosen_animal_ids: symbolize(%{bossie.id => "true",
                                             unchecked.id => "false"})}})
```

That works *provided* the user knows that all of the values have to be specified - relying on defaults doesn't do the right thing. Documentation might be sufficient for that? I'm pretty good at writing documentation. 

The second case (which I'm told was the original idea behind the checkbox tag) could be written by:

```elixir
        follow_form(has_animal_form,
          %{animals:
            %{chosen_animal_ids: ["true", "false"})}})
```

**Comment by boydm** [comment]
I am super glad you are thinking this through! My head is full of C code right now and I worry I wouldn't be able to give it proper attention like this right now.

**Comment by boydm** [comment]
If this is just a Documentation issue, that is a total win! I'd get that into the build right away...

**Comment by marick** [comment]
There could be something more ambitious that would detect the sort of error I made, but that will be a fair amount of work.


The documentation will be longer than should be in an API description. In the past, I've put those in the project's wiki: https://github.com/marick/midje/wiki

Alternately, I could put it on my new blog (but that runs the risk of it disappearing). I'd probably also want to put it on my [new blog](https://www.crustofcode.com/).

**Comment by marick** [comment]
Note that the sort of error I made is, I think, really easy to make. It's an [attractive nuisance](https://en.wikipedia.org/wiki/Attractive_nuisance_doctrine).

**Comment by boydm** [comment]
How about this: Do a write-up in the style of a blog. Heck. Post it as a blog.

But we'll also add it as a guide to the docs. Might add another one or two as well. That way it has some permanence. I did this sort of thing for Scenic. There is a nice mechanism to add "guides" to the docs, which are basically blog posts that sit beside the api docs.

**Comment by boydm** [comment]
If you do write up as a *.md file, then put it into a new top-level "/guides" directory, that would be easy for me to hook into the main documentation.

**Comment by marick** [comment]
I got a good start on the blog post, but then I let myself be tempted into coding. I think I have a good start on handling two features that `phoenix_integration` is missing:
1. The full complexity of the HTML `type="checkbox"` input, including the "`hidden` hack". 
2. Fully supporting the Phoenix's special-case handling of names like `animals[nicknames][]`. 

Before I go further, though, I should check if the direction I'm taking will produce a pull request you'd accept. Basically, I think I can produce a nested map that better represents the starting state of an HTML form. Then overriding values in `follow_form` will be less surprising.

So far, I've kept my new code separate from the existing code. The tests/examples are [here](https://github.com/marick/phoenix_integration/tree/master/test/details). The code is [here](https://github.com/marick/phoenix_integration/tree/master/lib/phoenix_integration/form). 

Since you're a fan of Elixir's `|>` operator for tests, you might be interested that I added my own `assert_fields` to the project, allowing tests like [this](https://github.com/marick/phoenix_integration/blob/master/test/details/tag_test.exs#L16). I've found that chaining assertions make it more clear what tests are actually about.

**Comment by boydm** [comment]
I'm open to this.

I really like assert_fields and am all in on that.

Don't have my head wrapped around build_tree and how to set up a test. Do you have a very simple real-world example around how I would test the form?

**Comment by marick** [comment]
What I'm thinking of is just a change to `Requests.build_form_data` that leaves its interface alone. So all of the existing tests, even those that call `test_build_form_data`, would be unchanged. Like this:

```elixir 
  ####### FROM
  defp build_form_data(form, form_action, fields) do
    form_data = get_form_data(form)
    # merge the data from the form and that provided by the test
    merge_grouped_fields(form_data, form_action, fields)
  end

  ####### TO
  defp build_form_data__2(form, form_action, user_tree) do
    form
    |> Tree.build_tree
    |> Tree.merge_user_tree(user_tree, form_action)
  end
```

However, it wouldn't be particularly hard to add two things:
1. Some warnings in cases where the HTML is something that Phoenix actually accepts but likely doesn't do what you want. For example, you have two non-array `inputs` with the same name (Phoenix throws away the first.) 
2. Multiple error messages for the user overrides instead of just one. 

That would look like this:

```elixir
  defp build_form_data__3(form, user_tree) do
    with(
      {:ok, initial_tree, warnings} <- Tree.build_tree(form),
      {:ok, result} <- Tree.merge_user_tree(initial_tree, user_tree)
    ) do
      log_form_warnings(warnings, form)
      result
    else
      {:error, errors} ->
        log_form_errors(errors, form)
        raise "Stopping."
    end
  end
```

Note that you wouldn't have to pass the form's action to the `build_form_data`, because it's readily available in the form.

**Comment by marick** [comment]
The checkbox tests now pass. For example, the code now correctly handles `type="hidden"` tags that supply default values:

```html
<form accept-charset="UTF-8" action="/form" id="with_hidden" method="post">
  <input name="animals[chosen][1]" type="hidden" value="false">
  <input name="animals[chosen][1]" type="checkbox" value="true">

  <input name="animals[chosen][2]" type="hidden" value="false">
  <input name="animals[chosen][2]" type="checkbox" value="true">
</form>
```

```elixir
    test "if not 'checked' by the test, the default (hidden) values are used",
      %{form_with_hidden: form} do
      
      %{animals: %{chosen: checked}} =
        test_build_form_data__2(form, @form_action, %{})
      assert checked == %{"1": "false", "2": "false"}
    end
```

**Comment by marick** [comment]
Closed by #34
